### PR TITLE
AuthenticationFailed and ReceivingToken events added.

### PR DESCRIPTION
**Pull request by DovydasNavickas**
First two events to agree if this is what we need.

### CODE REVIEW

**Comment by DovydasNavickas** [comment]
@PinpointTownes, what I've done is basically copied and adjusted code from https://github.com/aspnet/Security/tree/dev/src/Microsoft.AspNetCore.Authentication.JwtBearer
Because it's... Hard to think something else. They have all of the nitty-gritty stuff handled already, which makes coding everything myself kind of pointless.

If this is what we need - I can take https://github.com/aspnet-contrib/AspNet.Security.OAuth.Extensions/issues/2.

One thing that I'd like to discuss is whitespace configuration... But more on that in a different issue, to not pollute this one.

**Comment by PinpointTownes** [comment]
> If this is what we need - I can take #2.

That's, that's exactly what we need, thanks! :smile:

File: src/AspNet.Security.OAuth.Validation/Events/BaseValidationContext.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+﻿using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using System;
+
+namespace AspNet.Security.OAuth.Validation.Events
+{
+    public abstract class BaseValidationContext
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/Events/BaseValidationContext.cs]: This base class is not necessary, as aspnet/Security already offers a similar one that does the same thing: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/Events/BaseControlContext.cs#L8

File: src/AspNet.Security.OAuth.Validation/Events/IValidationEvents.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Threading.Tasks;
+
+namespace AspNet.Security.OAuth.Validation.Events
+{
+    public interface IValidationEvents
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/Events/IValidationEvents.cs]: `IValidationEvents` is too generic. Consider renaming it to `IOAuthValidationEvents`

File: src/AspNet.Security.OAuth.Validation/Events/ValidationEvents.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace AspNet.Security.OAuth.Validation.Events
+{
+    public class ValidationEvents : IValidationEvents
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/Events/ValidationEvents.cs]: `OAuthValidationEvents`

File: src/AspNet.Security.OAuth.Validation/Events/ValidationEvents.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        /// Invoked if exceptions are thrown during request processing. The exceptions will be re-thrown after this event unless suppressed.
+        /// </summary>
+        public Func<AuthenticationFailedContext, Task> OnAuthenticationFailed { get; set; } = context => Task.FromResult(0);
+
+        /// <summary>
+        /// Invoked when a protocol message is first received.
+        /// </summary>
+        public Func<ReceivingTokenContext, Task> OnReceivingToken { get; set; } = context => Task.FromResult(0);
+
+        public Task AuthenticationFailed(AuthenticationFailedContext context) => OnAuthenticationFailed(context);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/Events/ValidationEvents.cs]: `virtual`

File: src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs [hunk 5]
```diff
@@ -10,68 +10,127 @@
...
+                // Ensure that the authentication ticket is still valid.
+                if (ticket.Properties.ExpiresUtc.HasValue &&
+                    ticket.Properties.ExpiresUtc.Value < Options.SystemClock.UtcNow)
+                {
+                    return AuthenticateResult.Fail("Authentication failed because the access token was expired.");
+                }
+
+                return AuthenticateResult.Success(ticket);
             }
+            catch (Exception ex)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs]: nit: `ex` -> `exception`

File: src/AspNet.Security.OAuth.Validation/OAuthValidationOptions.cs [hunk 6]
```diff
@@ -31,5 +35,12 @@ public class OAuthValidationOptions : AuthenticationOptions {
         /// authenticated tickets received by the validation middleware.
         /// </summary>
         public ISecureDataFormat<AuthenticationTicket> TicketFormat { get; set; }
+
+        /// <summary>
+        /// The object provided by the application to process events raised by the validation authentication middleware.
+        /// The application may implement the interface fully, or it may create an instance of ValidationEvents
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationOptions.cs]: `OAuthValidationEvents`.

Consider using the `<see />` syntax.

File: src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs [hunk 7]
```diff
@@ -10,68 +10,127 @@
...
+            try
+            {
+                // Give application opportunity to find from a different location, adjust, or reject token
+                var receivingTokenContext = new ReceivingTokenContext(Context, Options);
 
-                return AuthenticateResult.Skip();
-            }
+                // event can set the token
+                await Options.Events.ReceivingToken(receivingTokenContext);
+                if (receivingTokenContext.HandledResponse)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs]: Should we have a test or two to ensure we never regress this path?

File: src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs [hunk 8]
```diff
@@ -10,68 +10,127 @@
...
+                }
 
-            // Ensure that the authorization header contains the mandatory "Bearer" scheme.
-            // See https://tools.ietf.org/html/rfc6750#section-2.1
-            if (!header.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase)) {
-                return AuthenticateResult.Fail("Authentication failed because an invalid scheme " +
-                                               "was used in the 'Authorization' header.");
-            }
+                // If application retrieved token from somewhere else, use that.
+                token = receivingTokenContext.Token;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs]: Same remark.

File: src/AspNet.Security.OAuth.Validation/Events/BaseValidationContext.cs (hunk 1 above)
Comment (DovydasNavickas) [inline src/AspNet.Security.OAuth.Validation/Events/BaseValidationContext.cs]: :+1:

File: src/AspNet.Security.OAuth.Validation/Events/IValidationEvents.cs (hunk 2 above)
Comment (DovydasNavickas) [inline src/AspNet.Security.OAuth.Validation/Events/IValidationEvents.cs]: I thought of that. It was just a fast-code (just as unhealthy as fast-food). I just wanted to know if it's what we need :+1:

Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/Events/IValidationEvents.cs]: Yep, you're definitely on the right track! :clap:

**Comment by PinpointTownes** [comment]
As announced on Gitter, the `vNext` branch will be removed and the ASP.NET Core middleware will be merged in the `dev` branch.

Please re-submit your PR against the `dev` branch :clap:

**Pull request closed without merging**